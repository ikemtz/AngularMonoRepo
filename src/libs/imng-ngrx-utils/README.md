# imng-ngrx-utils

This library was generated with [Nx](https://nx.dev).

## Running unit tests

Run `nx test imng-ngrx-utils` to execute the unit tests.
