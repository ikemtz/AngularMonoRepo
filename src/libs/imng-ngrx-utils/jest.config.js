module.exports = {
  name: 'imng-ngrx-utils',
  preset: '../../jest.config.js',
  coverageDirectory: '../../coverage/libs/imng-ngrx-utils',
  snapshotSerializers: [
    'jest-preset-angular/build/AngularNoNgAttributesSnapshotSerializer.js',
    'jest-preset-angular/build/AngularSnapshotSerializer.js',
    'jest-preset-angular/build/HTMLCommentSerializer.js',
  ],
};
