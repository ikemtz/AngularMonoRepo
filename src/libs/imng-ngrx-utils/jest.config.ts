import type { Config } from 'jest';

const config: Config = {
  displayName: 'imng-ngrx-utils',
  reporters: [
    'default',
    [
      'jest-junit',
      {
        outputDirectory: '.',
        outputName: './junit.imng-ngrx-utils.xml',
      },
    ],
  ],
  setupFilesAfterEnv: ['<rootDir>/src/test-setup.ts'],
  globals: {},
  coverageDirectory: '../../coverage/libs/imng-ngrx-utils',
  transform: {
    '^.+\\.(ts|mjs|js|html)$': [
      'jest-preset-angular',
      {
        tsconfig: '<rootDir>/tsconfig.spec.json',
        stringifyContentPathRegex: '\\.(html|svg)$',
      },
    ],
  },
  transformIgnorePatterns: ['node_modules/(?!.*\\.mjs$)'],
  snapshotSerializers: [
    'jest-preset-angular/build/serializers/no-ng-attributes',
    'jest-preset-angular/build/serializers/ng-snapshot',
    'jest-preset-angular/build/serializers/html-comment',
  ],
  preset: '../../jest.preset.js',
};

export default config;
