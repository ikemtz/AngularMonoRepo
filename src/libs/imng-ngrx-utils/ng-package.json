{
  "$schema": "../../node_modules/ng-packagr/ng-package.schema.json",
  "dest": "../../dist/libs/imng-ngrx-utils",
  "lib": {
    "entryFile": "src/lib/public_api.ts"
  },
  "deleteDestPath": true
}
