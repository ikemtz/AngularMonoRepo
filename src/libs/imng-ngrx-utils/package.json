{
  "name": "imng-ngrx-utils",
  "version": "0.0.4",
  "author": "Isaac Martinez <@ikemtz>",
  "license": "MIT",
  "peerDependencies": {
    "@angular/common": "^8.0.0",
    "@angular/core": "^8.0.0",
    "@progress/kendo-data-query": "^1.5.1",
    "@ngrx/store": "^8.0.1"
  },
  "devDependencies": {},
  "ngPackage": {
    "$schema": "./node_modules/ng-packagr/ng-package.schema.json",
    "lib": {
      "entryFile": "src/index.ts"
    },
    "dest": "../../dist/libs/imng-ngrx-utils"
  }
}
