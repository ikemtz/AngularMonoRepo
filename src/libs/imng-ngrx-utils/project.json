{
  "name": "imng-ngrx-utils",
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "projectType": "library",
  "sourceRoot": "libs/imng-ngrx-utils/src",
  "prefix": "imng",
  "tags": [],
  "targets": {
    "build": {
      "executor": "@nx/angular:package",
      "outputs": ["{workspaceRoot}/dist/libs/imng-ngrx-utils"],
      "dependsOn": [
        {
          "target": "build",
          "dependencies": true
        }
      ],
      "options": {
        "project": "libs/imng-ngrx-utils/ng-package.json"
      },
      "configurations": {
        "production": {
          "tsConfig": "libs/imng-ngrx-utils/tsconfig.lib.prod.json"
        },
        "development": {
          "tsConfig": "libs/imng-ngrx-utils/tsconfig.lib.json"
        }
      },
      "defaultConfiguration": "production"
    },
    "test": {
      "executor": "@nx/jest:jest",
      "outputs": ["{workspaceRoot}/coverage/libs/imng-ngrx-utils"],
      "options": {
        "jestConfig": "libs/imng-ngrx-utils/jest.config.ts"
      }
    },
    "lint": {
      "executor": "@nx/eslint:lint"
    }
  }
}
