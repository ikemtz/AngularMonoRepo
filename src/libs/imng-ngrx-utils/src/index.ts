export { Payload, createPayloadAction } from './lib/payload';
export { isTruthy, isFalsy } from './lib/isTruthy';
export { findAndModify } from './lib/find-and-modify';
export { findAndMerge } from './lib/find-and-merge';
export { matchAndSpread } from './lib/match-and-spread';
export { Subscriptions } from './lib/subscriptions';
export { Subscribable } from './lib/subscribable';
export { idType } from './lib/id-type';
