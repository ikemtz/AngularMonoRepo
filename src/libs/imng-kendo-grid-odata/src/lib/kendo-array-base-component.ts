import { DataStateChangeEvent, SortSettings, PageChangeEvent } from '@progress/kendo-angular-grid';
import { Input, ChangeDetectorRef, OnDestroy, Component } from '@angular/core';
import { process, State, CompositeFilterDescriptor, SortDescriptor } from '@progress/kendo-data-query';
import { ODataResult } from 'imng-kendo-odata';
import { BehaviorSubject } from 'rxjs';
import { Subscribable, Subscriptions } from 'imng-ngrx-utils';

// eslint-disable-next-line @angular-eslint/directive-class-suffix
@Component({ template: '' })
export abstract class KendoArrayBasedComponent<PARENT_ENTITY, LISTED_ENTITY> implements OnDestroy, Subscribable {
  public readonly allSubscriptions = Subscriptions.instance;

  @Input() public item?: PARENT_ENTITY;

  private _detail: LISTED_ENTITY[];
  @Input()
  public set detail(value: LISTED_ENTITY[]) {
    this._detail = value || [];
    this.gridData = process(this._detail, this.state);
    this.changeDetectorRef?.markForCheck();
  }

  /**
   * A properties enum to make kendo grid columns definitions type safe
   * {@example <kendo-grid-column [field]="props.FIELD_NAME">}
   */
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  public abstract readonly props: any; //NOSONAR
  public state: State = {
    skip: 0,
    take: 10,
  };

  public readonly sortSettings: SortSettings = {
    allowUnsort: true,
    mode: 'multiple',
  };

  private _gridData: ODataResult<LISTED_ENTITY> | LISTED_ENTITY[];
  public gridData$ = new BehaviorSubject<ODataResult<LISTED_ENTITY> | LISTED_ENTITY[]>({ data: [], total: 0 });
  public get gridData(): ODataResult<LISTED_ENTITY> | LISTED_ENTITY[] {
    return this._gridData;
  }
  public set gridData(value: ODataResult<LISTED_ENTITY> | LISTED_ENTITY[]) {
    this._gridData = value;
    this.gridData$.next(value);
    this.markForCheck();
  }

  constructor(public readonly changeDetectorRef: ChangeDetectorRef = null) {}

  public readonly markForCheck = (): void => this.changeDetectorRef?.markForCheck();

  public dataStateChange(state: DataStateChangeEvent | State): void {
    this.state = state;
    this.gridData = process(this._detail, this.state);
  }

  // eslint-disable-next-line @typescript-eslint/no-empty-function, @typescript-eslint/no-unused-vars
  public pageChange(t: PageChangeEvent): void {}

  // eslint-disable-next-line @typescript-eslint/no-empty-function, @typescript-eslint/no-unused-vars
  public filterChange(t: CompositeFilterDescriptor): void {}

  // eslint-disable-next-line @typescript-eslint/no-empty-function, @typescript-eslint/no-unused-vars
  public sortChange(t: SortDescriptor[]): void {}

  public ngOnDestroy(): void {
    this.allSubscriptions.unsubscribeAll();
  }
}
