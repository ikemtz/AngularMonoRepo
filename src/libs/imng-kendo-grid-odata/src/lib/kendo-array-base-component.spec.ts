import { ComponentFixture, TestBed } from '@angular/core/testing';
import { Component } from '@angular/core';
import { KendoArrayBasedComponent } from './kendo-array-base-component';
import { GridModule, GridComponent } from '@progress/kendo-angular-grid';
import { ImngArrayGridDirective } from './kendo-array-grid.directive';
import { By } from '@angular/platform-browser';
import { of } from 'rxjs';
import { Subscribable, Subscriptions } from 'imng-ngrx-utils';

const template = '<kendo-grid [imngArrayGrid]="this"><kendo-grid-column field="id"></kendo-grid-column></kendo-grid>';
describe('KendoODataComponentBase', () => {
  let component: KendoArrayGridTestComponent;
  let fixture: ComponentFixture<KendoArrayGridTestComponent>;

  beforeEach(() => {
    TestBed.configureTestingModule({
      declarations: [KendoArrayGridTestComponent, ImngArrayGridDirective],
      imports: [GridModule],

      providers: [],
    }).compileComponents();
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(KendoArrayGridTestComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should handle dataStateChange', () => {
    const dataStateChangeSpy = jest.spyOn(component, 'dataStateChange');
    const grid = fixture.debugElement.query(By.directive(GridComponent)).injector.get(GridComponent);
    grid.dataStateChange.emit({ take: 3, skip: 7, sort: [{ field: 'id', dir: 'desc' }] });
    expect(component.state).toStrictEqual({ take: 3, skip: 7, sort: [{ field: 'id', dir: 'desc' }] });
    expect((component as unknown as { _gridData; })._gridData).toMatchSnapshot();
    expect(dataStateChangeSpy).toBeCalledTimes(1);
  });

  it('should handle pageChange', () => {
    const pageChangeSpy = jest.spyOn(component, 'pageChange');
    const dataStateChangeSpy = jest.spyOn(component, 'dataStateChange');
    const grid = fixture.debugElement.query(By.directive(GridComponent)).injector.get(GridComponent);
    grid.pageChange.emit({ take: 2, skip: 4 });
    expect(component.state).toStrictEqual({ take: 2, skip: 4, group: [], sort: [], filter: undefined });
    expect((component as unknown as { _gridData; })._gridData).toMatchSnapshot();
    expect(pageChangeSpy).toBeCalledTimes(1);
    expect(dataStateChangeSpy).toBeCalledTimes(1);
  });

  it('should handle filterChange', () => {
    const grid = fixture.debugElement.query(By.directive(GridComponent)).injector.get(GridComponent);
    grid.filterChange.emit({ logic: 'and', filters: [{ field: 'id', operator: 'eq', value: 'ðŸ˜’ðŸ˜’' }] });
    expect(component.state).toStrictEqual({
      take: undefined, skip: 0, group: [], sort: [], filter:
        { logic: 'and', filters: [{ field: 'id', operator: 'eq', value: 'ðŸ˜’ðŸ˜’' }] }
    });
  });

  it('should handle sortChange', () => {
    const dataStateChangeSpy = jest.spyOn(component, 'dataStateChange');
    const sortChangeSpy = jest.spyOn(component, 'sortChange');
    const grid = fixture.debugElement.query(By.directive(GridComponent)).injector.get(GridComponent);
    grid.pageChange.emit({ take: 3, skip: 1 });
    grid.sortChange.emit([{ field: 'id', dir: 'asc' }]);
    expect(component.state).toStrictEqual({ filter: undefined, take: 3, skip: 1, group: [], sort: [{ field: 'id', dir: 'asc' }] });
    expect((component as unknown as { _gridData: never; })._gridData).toMatchSnapshot();
    expect(dataStateChangeSpy).toBeCalledTimes(2);
    expect(sortChangeSpy).toBeCalledTimes(1);
  });

  it('should destroy', () => {
    const imngDirective = fixture.debugElement.query(By.directive(ImngArrayGridDirective)).injector.get(ImngArrayGridDirective);
    fixture.destroy();
    (imngDirective as unknown as { allSubscriptions: Subscriptions; }).allSubscriptions.forEach((t) => expect(t.closed).toBeTruthy());
  });
});

@Component({
  selector: 'imng-test-component',
  template: template,
})
export class KendoArrayGridTestComponent
  // eslint-disable-next-line @typescript-eslint/ban-types
  extends KendoArrayBasedComponent<object, object>
  implements Subscribable {
  state = {};
  props = {};
  constructor() {
    super();
    this.detail = [{ id: 6 }, { id: 9 }, { id: 10 }, { id: 1 },
    { id: 2 }, { id: 3 }, { id: 4 }, { id: 5 }, { id: 7 }, { id: 8 },
    { id: 11 }, { id: 12 }, { id: 13 }, { id: 14 }, { id: 15 }, { id: 16 },
    { id: 17 }, { id: 18 }, { id: 19 }];
    this.allSubscriptions.push(of(123).subscribe());
  }
}
