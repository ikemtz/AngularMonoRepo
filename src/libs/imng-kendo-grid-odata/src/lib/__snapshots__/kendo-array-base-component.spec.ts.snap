// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KendoArrayBaseComponent should handle dataStateChange 1`] = `
Object {
  "data": Array [
    Object {
      "id": 12,
    },
    Object {
      "id": 11,
    },
    Object {
      "id": 10,
    },
  ],
  "total": 19,
}
`;

exports[`KendoArrayBaseComponent should handle pageChange 1`] = `
Object {
  "data": Array [
    Object {
      "id": 2,
    },
    Object {
      "id": 3,
    },
  ],
  "total": 19,
}
`;

exports[`KendoArrayBaseComponent should handle sortChange 1`] = `
Object {
  "data": Array [
    Object {
      "id": 2,
    },
    Object {
      "id": 3,
    },
    Object {
      "id": 4,
    },
  ],
  "total": 19,
}
`;
